### ISSUE TITLE
IPv6 forwarding no longer automatically enabled since firewalld v0.8.0

### ISSUE THREAD

**Issue by toreanderson** [firewalld/firewalld#683]
After upgrading firewalld from v0.7.0 to v0.8.0, I found that IPv6 forwarding no longer is enabled automatically, which broke all external IPv6 Internet connectivity from internal nodes after the first reboot. It remained broken until I manually issued `sysctl net/ipv6/conf/all/forwarding=1`.

The actual zone configuration did not change between the upgrades. Firewalld is acting in a traditional home gateway role, with an external `wan` zone facing the ISP and a `lan` zone facing the internal network. The `wan` zone has `masquerade: yes`, which is necessary due the use of private IPv4 addresses in the `lan` zone.

I believe it is likely that this is an unintentional side-effect of commit 88e13653686e4b72b4964e41564c70ca0096e0a9. It would appear that the code that enables the `forwarding` sysctl is dependent on `masquerade: yes`, and 88e13653686e4b72b4964e41564c70ca0096e0a9 makes `masquerade` apply only to IPv4, so the call that enables the `forwarding` sysctl is no longer reached for the IPv6 protocol.

NAT is generally not used with IPv6, so it makes perfect sense to make `masquerade` IPv4-only. However it remains required to enable the `forwarding` sysctl for both protocols in order to get forwarding between zones.

**Comment by erig0** [comment]
This is not a bug. Docs (man `firewall-cmd`) say that `--add-masquerade` only applies to IPv4. To enable IPv6 masquerade and therefore forwarding sysctl you need to use a rich rule:
```
# sysctl -a |grep net.ipv6.conf.all.forwarding
net.ipv6.conf.all.forwarding = 0
[# firewall-cmd --add-rich-rule='rule family=ipv6 masquerade'
success
# sysctl -a |grep net.ipv6.conf.all.forwarding
net.ipv6.conf.all.forwarding = 1
```

---

```
[--permanent] [--zone=zone] --add-masquerade [--timeout=timeval]
           Enable IPv4 masquerade for zone. If zone is omitted, default zone will be used. If a timeout is supplied, masquerading will be active for the specified
           amount of time.  timeval is either a number (of seconds) or number followed by one of characters s (seconds), m (minutes), h (hours), for example 20m or 1h.
           Masquerading is useful if the machine is a router and machines connected over an interface in another zone should be able to use the first connection.

           The --timeout option is not combinable with the --permanent option.

           For IPv6 masquerading, please use the rich language.

           Note: IP forwarding will be implicitly enabled.
```

**Comment by toreanderson** [comment]
I think you misunderstood me. I do not want IPv6 masquerading, just plain forwarding. So the fact that masquerading does not get enabled for IPv6 is the desired behaviour and I agree that it was right to change this behaviour from 0.7.0.

The bug is that control of the forwarding sysctl appears to be dependent on masquerading being enabled, which does not really make sense. It is quite possible to perform forwarding between zones without masquerading. This can be true for both IPv4 and IPv6, but is almost always the case for IPv6.

However, due to the inappropriate "forwarding dependent on masquerading" logic, IPv6 forwarding no longer gets enabled by firewalld on startup. This means you end up in an inconsistent state where the IPv4 forwarding sysctl is managed by firewalld (assuming masquerade is enabled) while the IPv6 forwarding sysctl must be managed by something else.

However, if you do not use masquerading with IPv4, you (probably) end up in a situation where neither of the forwarding sysctls are managed by firewalld.

**Comment by erig0** [comment]
> I think you misunderstood me. I do not want IPv6 masquerading, just plain forwarding. So the fact that masquerading does not get enabled for IPv6 is the desired behaviour and I agree that it was right to change this behaviour from 0.7.0.
> 
> The bug is that control of the forwarding sysctl appears to be dependent on masquerading being enabled, which does not really make sense.

It does make sense. If you enable masquerading then you obviously need forwarding as well.

> It is quite possible to perform forwarding between zones without masquerading. This can be true for both IPv4 and IPv6, but is almost always the case for IPv6.

Right. But at the moment firewalld does _not_ support filtering between zones, so why would we ever enable forwarding? How would we even know to enable it?

> However, due to the inappropriate "forwarding dependent on masquerading" logic, IPv6 forwarding no longer gets enabled by firewalld on startup. This means you end up in an inconsistent state where the IPv4 forwarding sysctl is managed by firewalld (assuming masquerade is enabled) while the IPv6 forwarding sysctl must be managed by something else.

Firewalld only ever enables forwarding if needed (masquerade). It never disables forwarding. So it shouldn't conflict with out-of-band settings.

> However, if you do not use masquerading with IPv4, you (probably) end up in a situation where neither of the forwarding sysctls are managed by firewalld.

I don't see the problem with that.

**Comment by toreanderson** [comment]
> Right. But at the moment firewalld does _not_ support filtering between zones, so why would we ever enable forwarding? 

Say what? Isn't the entire point of a zone-based firewall to filter traffic between zones? (A quick Google search for *zone based firewall* seems to agree.)

In any case, I just double-checked and confirmed that firewalld *does* indeed filter traffic between zones:

If I ping from a host in the `lan` zone (target `ACCEPT`) to a host in the `wan` zone  the packet gets to its destination just fine and so does the replies.

However, if I ping from a host in the `wan` zone (target `%%REJECT%%`) to a host in the `lan`, the packet does *not* reach its destination, instead it is replied to by the firewalld node with an ICMP `Destination unreachable: Administratively prohibited` error.

The firewall rules on this node are exclusively managed by firewalld. Looking at `nft list ruleset` I can clearly see it has created a `filter_FORWARD` chain in the `firewalld` table, so firewalld clearly does concern itself with filtering forwarded traffic.

> How would we even know to enable it?

Something like `if ($number_of_active_zones > 1) { enable_forwarding(); }` would be logical to me.

Alternatively you could enable forwarding the moment you determine that a `type filter hook forward` chain should be created. Without forwarding being enabled the rules in such chains become completely meaningless no-ops, after all.

**Comment by erig0** [comment]
> > Right. But at the moment firewalld does _not_ support filtering between zones, so why would we ever enable forwarding?
> 
> Say what? Isn't the entire point of a zone-based firewall to filter traffic between zones? (A quick Google search for _zone based firewall_ seems to agree.)
> 
> In any case, I just double-checked and confirmed that firewalld _does_ indeed filter traffic between zones:
> 
> If I ping from a host in the `lan` zone (target `ACCEPT`) to a host in the `wan` zone the packet gets to its destination just fine and so does the replies.
> 
> However, if I ping from a host in the `wan` zone (target `%%REJECT%%`) to a host in the `lan`, the packet does _not_ reach its destination, instead it is replied to by the firewalld node with an ICMP `Destination unreachable: Administratively prohibited` error.
> 
> The firewall rules on this node are exclusively managed by firewalld. Looking at `nft list ruleset` I can clearly see it has created a `filter_FORWARD` chain in the `firewalld` table, so firewalld clearly does concern itself with filtering forwarded traffic.

Let me rephrase.. firewalld allows forward between zones and those zones have a `--set-target` that is a catch-all. But you can't add a service, port, rich rule, etc to a zone and have it affect forwarded traffic. The latter is what I was referring to.

> > How would we even know to enable it?
> 
> Something like `if ($number_of_active_zones > 1) { enable_forwarding(); }` would be logical to me.

I disagree. Most firewalld users don't need forwarding as it's often deployed as an end station firewall.

> Alternatively you could enable forwarding the moment you determine that a `type filter hook forward` chain should be created. Without forwarding being enabled the rules in such chains become completely meaningless no-ops, after all.

As I said above. For the most part firewalld doesn't support adding rules to forward chains. There are a couple of exceptions:
  - masquerade
  - ICMP block
  - `--set-target`
